### ISSUE TITLE
IPTV AU channel 7 hangs at start of ad break

### ISSUE THREAD

**Issue by SoreTissue** [matthuisman/slyguy.addons#305]
Hi Matt

After channel 7 stopped working about a week ago where it was reported as Channel 7 stream links not working #298 , you applied a fix where channel 7 works but it seems to get stuck on when an ad break starts. Similar to the issue reported for Pluto TV.

I've tried with "use Inputstream HLS for Live Streams" turned off in IPTV AU settings > Playback but it still occurs. It's only an issue with Channel 7 channels and only after the fix was applied mentioned in above issue #298 which also added more channels (really enjoying the Bathurst Throwback)

Thanks for your efforts btw

**Comment by matthuisman** [comment]
Out of my hands unfortunately :( their new streams use server side ad injection (x-discontuity) which doesn't seem to be supported in kodis ffmpeg or inputstream adaptive

**Comment by SoreTissue** [comment]
Ok. Thx for the response. I was making a log file and was noticing more local ads and how it was pausing on each local ad. In the log it was only pausing for a few seconds, in the past it's got stuck and you have to restart channel. 

Thanks again for your efforts and swift response

**Comment by matthuisman** [comment]
Get me a Kodi debug log with the stopping at ads occuring and I can create
an issue for inputstream adaptive

**Comment by SoreTissue** [comment]
[kodi.log](https://github.com/matthuisman/slyguy.addons/files/10048382/kodi.log)

I ran debug on channel 7Flix through 2 ad breaks but the file was too large to paste to https://paste.kodi.tv/.

Hope it's ok to upload the file here. If not let me know a website that accepts larger pastes (2.58mb) or how to get a smaller debug log

**Comment by matthuisman** [comment]
thanks. So, did playback actually stop in that log?

**Comment by SoreTissue** [comment]
Playback paused for a few seconds at the start of every ad break. It seemed the first 2 ads were local and each had to load, then it reverted to a more generic ad stream. For the past week it used to hang with a percentage sign on screen but it would never restart itself. This was happening with 2 other people as well that I know personally which spurred me to log an issue here. It just so happened when I ran the debug log, it wasn't getting stuck for any more than a few seconds.

Info on debug file:
turned on debug, restarted kodi, went to 7flix and waited for 2 ad breaks. At the start of each ad break, it paused for around 5 seconds, It did this for the first 2 ads in each ad break. I presume it was due to those first 2 ads were local to my region which I haven't noticed before. After 2 ad breaks, I stopped stream and exited Kodi.

Yesterday I tried android tv addon called PVR Live (tried many including TVirl but like how this one still auto-updates epg but at the cost of logos). It uses the Live Channels app installed. I've installed this on 2 devices at different houses. I used your m3u url (https://i.mjh.nz/au/{REGION}/kodi-tv.m3u8) to add channels and 7flix does not hang at all at ad breaks. So in with my limited knowledge it seems like it's a Kodi issue as you stated and not an issue with the m3u.

If you need further logs or more specific way to run a debug log, I'm more than happy to.

**Comment by matthuisman** [comment]
Last question, are you any using any sort of refresh rate switching setting in Kodi?

**Comment by matthuisman** [comment]
have opened issue here: https://github.com/xbmc/inputstream.adaptive/issues/1071

**Comment by SoreTissue** [comment]
I have Settings/Player/Videos/adjust display refresh rate "off".  I run Kodi mostly on default settings and was a clean install a few weeks ago.

The only thing I've changed in Settings/Player is in Language, preferred audio language to English and Preferred Subtitle language to Forced only. 

No settings have been in changed in Settings/System (other than add-ons Unkown sources)

No fancy setup, Nvidia Shield Direct into TV, default skin, no local library, only 2 addons installed with its program addons plus youtube,

Basically, the same setup on the other 2 people I know that had this issue too. I have spoken to them since and they definitely said it was getting stuck for them too. Long enough to leave the room and do things and come back and still stuck on the loading wheel. 

Thanks again for your efforts

**Comment by matthuisman** [comment]
thanks for the reply. Are they on Android as well?
have you seen it do it on any non-android devices? eg. Windows or Rasberry Pi etc

**Comment by SoreTissue** [comment]
Yes 2 Shields and CCwGT

I have not observed it on other devices but I'll install it on pc and only install IPTV AU and come back with an answer.

**Comment by SoreTissue** [comment]
I tested both Kodi 19.4 and Kodi 20 latest nightly on Windows 10.

On a clean install:
I only changed Settings/Interface/Regional to my Region
I only installed IPTV AU and the dependencies it asks for
IPTV AU Settings I turned it to my Region, changed Playback Quality to Best and chose IPTV AU/Settings/Live TV&Epg/Setup IPTV Merge


I set up Shield and PC side by side and watched 7Flix to observe if there was any difference in behaviour.

Both 19.4 and 20 paused for 3-5 for the local ads but on Windows 10 it didn't show the Loading Wheel. Other than the loading wheel, behaviour between Windows and Android seemed very similiar. 

I then closed Kodi on the Shield and Tried PVR Live android tv app and also Sparkle TV (made by same guy) to observe any difference between the Windows Kodi version and the Shield. I used https://i.mjh.nz/au/{REGION}/kodi-tv.m3u8 In both those apps. 

Both Android TV apps ran 7Flix flawlessly. The apps showed the Localised ads but it doesn't stall at all at any time through the ad break or show. 

Whilst on the Windows Kodi it was pausing at the start, and sometimes end, on the localised ads. Usually 2 at the start of the ad break. Sometimes only a few seconds, even one time actually played flawlessy, but in the end it stalled for so long (longer than a minute) I got up and stopped the channel and called it a night on the testing.

**Comment by matthuisman** [comment]
That's really good testing. Thank you. Hopefully w have enough for one of the IA devs to find the issue :)

**Comment by Thomo351** [comment]
Hi Matt
After testing some settings yesterday 7 mate worked fine through the adds, a little bit slow loading initially but changing channel time depended on free to air or Foxtel selection

Kodi Build 18.5 Git on Nvidia Shield

My settings
Kodi Player : Adjust refresh / Always
Enable HQ 20%
Media codecs both On
IPTVAU : Melbourne : Playback/Bypass : HLS / ON
Input adaptive stream : Ignore Display / On : HDCP Status / On
IPTV Merge : Playback/ Best : HTTP / On : HLS / On

Hope that is some help
Peter Thomas

**Comment by Thomo351** [comment]
As For Live Channels I use CosmiDVR and Free to air and Pluto using your'e mu38 links are working fine

**Comment by matthuisman** [comment]
Kodi 18.5? Is that typo?

**Comment by Thomo351** [comment]
No Matt that what is says in settings, everything has been working so I havent upgraded it

**Comment by Thomo351** [comment]
7Mate been running again today for 3 hours without problem

**Comment by matthuisman** [comment]
have you had issues on 18.5 before changing settings?
I think IA / kodi doesn't do the mutil-chapter thing in 18.5
Which is my feeling may be the cause

**Comment by Thomo351** [comment]
All was working good with free to air and Foxtel before channel 7 changed its streams which you have explained
Foxtel has always been OK and the settings I changed have now have fixed the 7mate adds problems (for now ?) Foxtel has always been OK. Im not sure what the settings were before I had the problem but I just installed the addons with the default settings.
Im new to this but I assume by IA /Kodi u mean Input Adaptive my version is 2.48

**Comment by matthuisman** [comment]
Change IPTVAU : Melbourne : Playback/Bypass : HLS / ON to Playback/Best
should help speed up initial load a bit

Be interested to see if you can find which single setting introduces the advert bug

**Comment by glennguy** [comment]
Hey guys, I think it might be better for me to comment here to get the whole audience...

Haven't been able to see what's going on with the discontinuities - they're not happening while I'm watching 7flix right now. Can still tell that the ads affect the playlists as the segment du

Still no discos after typing this but maybe it depends on the program content to whether their packager inserts the discos?

Something I did notice - because of the potential for a short segment leading into an ad break (or out of an ad break) there is also potential that Kodi will run over the live edge and have to wait for a new segment to come in the next refresh - this happened to me while watching. Only paused for about a second.

What happens though when combined with a discontinuity? The long standing bug that I still haven't been able to document properly. If a new period/chapter starts and the buffer can't get filled quick enough, a condition happens where Kodi/VideoPlayer ends up not having the correct playing time and thinks it's out of sync - by a value that is equal to how long the stream has been playing for. If the stream has been playing for 10 minutes and this happens, you'll be waiting 10 minutes for playback to resume.

Potential workaround for now:
Try rewinding the stream 20 or 30 seconds when you start playback.

**Comment by Thomo351** [comment]
I will see what I can do, it took awhile to get it going, waiting for the adds etc, my luck it will be the last setting I try

 

Cheers thomo

**Comment by CHJ85** [comment]
If you're interested, I found a workaround. By replacing SamsungTV with RokuChannel, which gets rid of the ad break.

Replace:
`?deviceType=samsung-tvplus&deviceMake=samsung&deviceModel=samsung&deviceVersion=unknown&appVersion=unknown&deviceLat=0&deviceLon=0&deviceDNT=%7BTARGETOPT%7D&deviceId=%7BPSID%7D&advertisingId=%7BPSID%7D&us_privacy=1YNY&samsung_app_domain=%7BAPP_DOMAIN%7D&samsung_app_name=%7BAPP_NAME%7D&profileLimit=&profileFloor=&embedPartner=samsung-tvplus`
With:
`?deviceId=channel&deviceModel=web&deviceVersion=1.0&appVersion=1.0&deviceType=rokuChannel&deviceMake=rokuChannel&deviceDNT=1&advertisingId=channel&embedPartner=rokuChannel&appName=rokuchannel&is_lat=1&bmodel=bm1&content=channel&platform=web&tags=ROKU_CONTENT_TAGS&coppa=false&content_type=livefeed&rdid=channel&genre=ROKU_ADS_CONTENT_GENRE&content_rating=ROKU_ADS_CONTENT_RATING&studio_id=viacom&channel_id=channel`

**Comment by matthuisman** [comment]
This is about 7plus ,:)

**Comment by matthuisman** [comment]
@Thomo351 , I suspect the inputstream hls setting to off will be the only thing that changes whether it works or not.

Odd that IA in Kodi 18 doesn't have issues with ads but 19 does. Makes me think it's Kodi issue with how Kodi handles chapter changes

**Comment by CHJ85** [comment]
> This is about 7plus ,:)
Oh. The addon 7Plus.
My bad. 😀
Reason I thought it was about the Pluto TV channel is because I saw another issue where somebody had trouble with ads crashing the channel.

**Comment by Thomo351** [comment]
Yea confirmed IPTVAU  HLS ON :  OK / OFF : it plays up

**Comment by matthuisman** [comment]
@glennguy I feel like kodis chapter changing stuff is cause? You added x-disc support in 18 and that works good. Then in 19, they started showing as chapters and the issues seem to have started. I think I've asked you before - but there is no way we can not tell Kodi about the chapters? Just one continuous stream. The chapters aren't useful anyway and can be annoying when used to navigating with skip 10s or whatever

**Comment by glennguy** [comment]
> but there is no way we can not tell Kodi about the chapters?

Two ways I guess...
* Remove disco tags
* introduce a setting to make IA ignore them

I'm not sure either is really good though, more of a hack. If the player actually needs to be reinitialized then problems will happen, like if ads are in a different resolution or frame rate to the main content. Also if there is a discontinuity in PTS there would probably be sync problems too (in fact I believe the bug I mentioned above has to do with Kodi either not setting PTS or a race causing PTS to be set back to 0).

**Comment by matthuisman** [comment]
Hmm, I wonder if IPTV simple / Kodi Pvr epg is the cause.

Has anyone ,on Kodi 19+ confirmed the same issues with adverts when playing the 7 channels directly in the addon and not via kodis tv menu?

**Comment by liam8888999** [comment]
I have had problems in the past on kodi 19 when playing channel 7 from directly in the addon. I think it all relates to ffmpeg not supporting discontinuities

**Comment by matthuisman** [comment]
there is a new PR against Inputstream Adaptive that should solve the advert issue on 7plus & pluto.

PR:
[https://github.com/xbmc/inputstream.adaptive/pull/1431](https://forums.whirlpool.net.au/go?https%3A%2F%2Fgithub.com%2Fxbmc%2Finputstream.adaptive%2Fpull%2F1431)

IA builds can be found here:
[https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-1431/1/artifacts](https://forums.whirlpool.net.au/go?https%3A%2F%2Fjenkins.kodi.tv%2Fblue%2Forganizations%2Fjenkins%2Fxbmc%252Finputstream.adaptive%2Fdetail%2FPR-1431%2F1%2Fartifacts)

You will need Kodi 21 Beta 2
